CPP: Add mentoring notes for space-age (#1551)
exercism/website-copy: tracks/cpp/exercises/space-age/mentoring.md

### Reasonable solutions
```cpp
#if !defined(_SPACE_AGE_H_)

> #define _SPACE_AGE_H_

Author: Didn't know that. Changed.

> #define _SPACE_AGE_H_

Reviewer: @MartinDelille Include guards vs. `#pragma once` still debated in the C++ community (see [StackOverflow](https://stackoverflow.com/a/34884735/1447760)).  
But [SF.8 of the C++ Core Guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf8-use-include-guards-for-all-h-files) takes a stand against `#pragma once` and recommends that you should always use `#include` guards in your header files.

#include <cstdint>

namespace space_age
{
    class space_age final
    {
    public:

>         explicit space_age(const uint64_t seconds);

Reviewer 2: I'd like to know more about the `explicit` constructor. Any resource or info?

>         explicit space_age(const uint64_t seconds);

Reviewer: [cppreference](https://en.cppreference.com/w/cpp/language/explicit) / [C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-explicit) / [Jason Turner's best practices](https://github.com/lefticus/cppbestpractices/blob/master/03-Style.md#avoid-implicit-conversions)

>         explicit space_age(const uint64_t seconds);

Author: @siebenschlaefer Can please you give me a referce for the namespace?

>         explicit space_age(const uint64_t seconds);

Reviewer: @elyashiv Sure. §17.6.1.2.4 [headers] in the C++14 specification: "In the C ++ standard library, however, the declarations (except for names which are defined as macros in C) are within namespace scope (3.3.6) of the namespace std." ([§16.5.1.2.5](http://eel.is/c%2B%2Bdraft/headers#5) in the current working draft)

        JUPITER = 374355660,
        SATURN = 929292360,
        URANUS = 2651370000,
        NEPTUNE = 5200418600,
    };

    space_age::space_age(const uint64_t seconds)
    : _seconds(seconds)
    { }

    uint64_t space_age::seconds() const
    {
        return _seconds;
    }

    double space_age::on_planet(const planet p) const
    {
        return _seconds * 1.0 / static_cast<uint64_t>(p);
    }

    double space_age::on_earth() const
    {
        return on_planet(planet::EARTH);
    }

    double space_age::on_mercury() const
    {
        return on_planet(planet::MERCURY);
    }

    double space_age::on_venus() const
    {
        return on_planet(planet::VENUS);
    }

    double space_age::on_mars() const
    {
        return on_planet(planet::MARS);
    }

    double space_age::on_jupiter() const
    {
        return on_planet(planet::JUPITER);
    }

    double space_age::on_saturn() const
    {
        return on_planet(planet::SATURN);
    }

    double space_age::on_uranus() const
    {
        return on_planet(planet::URANUS);
    }

    double space_age::on_neptune() const
    {
        return on_planet(planet::NEPTUNE);
    }
}
```
### Common suggestions

>  * Save seconds as a const in the class

Author: You are right. I'm just used to manage resources in a class, so I just make a non-copyable class without thinking unless I have a reason...

> ### Reasonable solutions

Reviewer 3: ```suggestion
### Reasonable solutions

```

```cpp
#if !defined(SPACE_AGE_H)
#define SPACE_AGE_H

#include <cstdint>

namespace space_age
{
    class space_age final
    {
    public:
        explicit space_age(const std::int64_t seconds);
        
        std::int64_t seconds() const;
        double on_earth() const;
        double on_mercury() const;
        double on_venus() const;
        double on_mars() const;
        double on_jupiter() const;
        double on_saturn() const;
        double on_uranus() const;
        double on_neptune() const;

    private:
        enum class planet : std::int64_t;
        double on_planet(const planet p) const;
        std::int64_t _seconds;
    };
}

#endif // SPACE_AGE_H
```
```cpp
#include "space_age.h"

namespace space_age
{
    enum class space_age::planet : std::int64_t
    {
        Earth = 31557600,
        Mercury = 7600544,
        Venus = 19414149,
        Mars = 59355036,
        Jupiter = 374355660,
        Saturn = 929292360,
        Uranus = 2651370000,

>         Neptune = 5200418600,

Reviewer 4: Should we expect users to convert everything to seconds or stick with units of "Earth years"? Or does it even matter?

Author: I don't think it matters.
One think that can be improved here is adding a unit to this values - seconds in this case.
I do wander if there is a way to use std::chrono for this, as it makes the units nice and clear, and allows converting easily.

    };

    space_age::space_age(const std::int64_t seconds)
    : _seconds(seconds)
    { }

    std::int64_t space_age::seconds() const
    {
        return _seconds;
    }

    double space_age::on_planet(const planet p) const
    {

>         return _seconds * 1.0 / static_cast<std::int64_t>(p);

Reviewer 4: ```suggestion
        return static_cast<double>(_seconds) / static_cast<std::int64_t>(p);
```
Seems more explicit than promoting implicitly through a multiply by 1.0.

    }

    double space_age::on_earth() const
    {
        return on_planet(planet::Earth);
    }

    double space_age::on_mercury() const
    {
        return on_planet(planet::Mercury);
    }

    double space_age::on_venus() const
    {
        return on_planet(planet::Venus);
    }

    double space_age::on_mars() const
    {
        return on_planet(planet::Mars);
    }

    double space_age::on_jupiter() const
    {
        return on_planet(planet::Jupiter);
    }

    double space_age::on_saturn() const
    {
        return on_planet(planet::Saturn);
    }

    double space_age::on_uranus() const
    {
        return on_planet(planet::Uranus);
    }

    double space_age::on_neptune() const
    {
        return on_planet(planet::Neptune);
    }
}
```

> ### Common suggestions

Reviewer 3: ```suggestion
### Common suggestions

```

 * Make sure the data type is big enough to store a 33 bit value (needed for neptune)
 * Make sure the data type is consistent (constructor, class member and return type from `seconds()`)
 * Constants should be static.
 * Arguments with names that indicate the units used.
 * Use `constexpr` for the functions.

> ### Talking points

Reviewer 3: ```suggestion
### Talking points

```

 * `explicit` constructor so you don't accidently do a type casting. [examples](https://en.cppreference.com/w/cpp/language/explicit).

>  * Prefer intialization over assinment. [Good explanation](https://stackoverflow.com/a/7350783)

Reviewer 4: assignment

 * Being explicit about units.

>  * `signed` vs. `unsigned`.
>  ** `unsigned` can give some suprising results. [example](isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es102-use-signed-types-for-arithmetic)

Reviewer 4: ```suggestion
 * `signed` vs. `unsigned`.
   * `unsigned` can give some surprising results. [example](isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es102-use-signed-types-for-arithmetic)
```

 * `double` vs. `float`

>  * Implementation in header:
>  ** This will turn the functions to be implicitly inline, and allow the compiler to do some optimizations on these very small functions.
>  ** This also means that every change in the implementation will require a recompile, and will create longer compile time.

Reviewer 4: ```suggestion
 * Implementation in header:
   * This will turn the functions to be implicitly inline, and allow the compiler to do some optimizations on these very small functions.
   * This also means that every change in the implementation will require a recompile, and will create longer compile time.
```

>  * `const` on seconds
>  ** We don't want to change it, so we can declare it const so we don't accidenlty do something stupid.
>  ** This won't allow us to use the assignment operator. Not needed for the exercise, but might be unintented.

Reviewer 4: ```suggestion
 * `const` on seconds
   * We don't want to change it, so we can declare it const so we don't accidentally do something stupid.
   * This won't allow us to use the assignment operator. Not needed for the exercise, but might be unintended.
```
